File: src/main/java/gargoyle/netsquares/util/Maps.java
package gargoyle.netsquares.util;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import java.util.function.Function;
import java.util.function.Supplier;
import java.util.stream.Collectors;

@SuppressWarnings({"", "unchecked", "Duplicates"})
public final class Maps {
    private Maps() {
    }

    private static @NotNull <K, V> Map<K, V> fill(@NotNull Map<K, V> map, Object... data) {
        for (int i = 0; i < data.length; i += 2) {
            map.put((K) data[i], (V) data[i + 1]);
        }
        return map;
    }

    public static <K, V> K keyForValue(Map<K, V> map, V value) {
        return map.entrySet().stream().filter(e -> Objects.equals(e.getValue(), value)).findFirst().map(Map.Entry::getKey).orElse(null);
    }

    public static <K, V> K keyForValueMultiMap(Map<K, List<V>> map, V value) {
        return map.entrySet().stream().filter(e -> e.getValue().contains(value)).findFirst().map(Map.Entry::getKey).orElse(null);
    }

    public static <K, V> K keyForValueMultiMapset(Map<K, Set<V>> map, V value) {
        return map.entrySet().stream().filter(e -> e.getValue().contains(value)).findFirst().map(Map.Entry::getKey).orElse(null);
    }

    public static <K, V> List<K> keysForValue(Map<K, V> map, V value) {
        return map.entrySet().stream().filter(e -> Objects.equals(e.getValue(), value)).map(Map.Entry::getKey).collect(Collectors.toList());
    }

    public static <K, V> List<K> keysForValueMultiMap(Map<K, List<V>> map, V value) {
        return map.entrySet().stream().filter(e -> e.getValue().contains(value)).map(Map.Entry::getKey).collect(Collectors.toList());
    }

    public static <K, V> List<K> keysForValueMultiMapset(Map<K, Set<V>> map, V value) {
        return map.entrySet().stream().filter(e -> e.getValue().contains(value)).map(Map.Entry::getKey).collect(Collectors.toList());
    }

    public static <K, V> Map<K, V> map(Object... data) {
        return Collections.unmodifiableMap(fill(new HashMap<>(), data));
    }

    public static @NotNull <K, V> Map<K, V> mapFrom(Supplier<Map<K, V>> source, Object... data) {
        return fill(source.get(), data);
    }

    public static <K, V> Map<K, List<V>> multiMap(@NotNull Class<K> keyType, @NotNull Class<V> valueType, Object... data) {
        //noinspection Convert2MethodRef
        return multiMapFrom(keyType, valueType, () -> new HashMap<>(), () -> new ArrayList<>(), data);
    }

    public static @NotNull <K, V> Map<K, List<V>> multiMapAdd(@NotNull Map<K, List<V>> map, K key, V value) {
        return multiMapAdd(ArrayList::new, map, key, value);
    }

    private static @NotNull <K, V> Map<K, List<V>> multiMapAdd(@NotNull Supplier<List<V>> listSource, Map<K, List<V>> map, K key, V value) {
        map.computeIfAbsent(key, k -> listSource.get()).add(value);
        return map;
    }

    @SafeVarargs
    public static @NotNull <K, V> Map<K, List<V>> multiMapAddAll(@NotNull Map<K, List<V>> map, K key, V... values) {
        return multiMapAddAll(ArrayList::new, map, key, values);
    }

    @SafeVarargs
    private static @NotNull <K, V> Map<K, List<V>> multiMapAddAll(@NotNull Supplier<List<V>> listSource, Map<K, List<V>> map, K key, V... values) {
        map.computeIfAbsent(key, k -> listSource.get()).addAll(Arrays.asList(values));
        return map;
    }

    private static <K, V> Map<K, List<V>> multiMapFrom(@NotNull Class<K> keyType, @NotNull Class<V> valueType, Supplier<Map<K, List<V>>> mapSource, @NotNull Supplier<List<V>> listSource, Object... data) {
        Map<K, List<V>> map = mapSource.get();
        K key = null;
        for (Object el : data) {
            if (keyType.isInstance(el)) {
                key = (K) el;
            } else if (valueType.isInstance(el)) {
                map.computeIfAbsent(key, k -> listSource.get()).add((V) el);
            } else {
                throw new IllegalArgumentException();
            }
        }
        return map;
    }

    public static <K, V> Map<K, Set<V>> multiMapset(@NotNull Class<K> keyType, @NotNull Class<V> valueType, Object... data) {
        //noinspection Convert2MethodRef
        return multiMapsetFrom(keyType, valueType, () -> new HashMap<>(), () -> new HashSet<>(), data);
    }

    public static @NotNull <K, V> Map<K, Set<V>> multiMapsetAdd(@NotNull Map<K, Set<V>> map, K key, V value) {
        return multiMapsetAdd(HashSet::new, map, key, value);
    }

    private static @NotNull <K, V> Map<K, Set<V>> multiMapsetAdd(@NotNull Supplier<Set<V>> listSource, Map<K, Set<V>> map, K key, V value) {
        map.computeIfAbsent(key, k -> listSource.get()).add(value);
        return map;
    }

    @SafeVarargs
    public static @NotNull <K, V> Map<K, Set<V>> multiMapsetAddAll(@NotNull Map<K, Set<V>> map, K key, V... values) {
        return multiMapsetAddAll(HashSet::new, map, key, values);
    }

    @SafeVarargs
    private static @NotNull <K, V> Map<K, Set<V>> multiMapsetAddAll(@NotNull Supplier<Set<V>> listSource, Map<K, Set<V>> map, K key, V... values) {
        map.computeIfAbsent(key, k -> listSource.get()).addAll(Arrays.asList(values));
        return map;
    }

    private static <K, V> Map<K, Set<V>> multiMapsetFrom(@NotNull Class<K> keyType, @NotNull Class<V> valueType, Supplier<Map<K, Set<V>>> mapSource, @NotNull Supplier<Set<V>> listSource, Object... data) {
        Map<K, Set<V>> map = mapSource.get();
        K key = null;
        for (Object el : data) {
            if (keyType.isInstance(el)) {
                key = (K) el;
            } else if (valueType.isInstance(el)) {
                map.computeIfAbsent(key, k -> listSource.get()).add((V) el);
            } else {
                throw new IllegalArgumentException();
            }
        }
        return map;
    }

    @SuppressWarnings("TypeMayBeWeakened")
    public static <S, T> List<T> transform(List<S> source, Function<S, T> mapper) {
        return Collections.unmodifiableList(source.stream().map(mapper).collect(Collectors.toList()));
    }
}

File: src/main/kotlin/gargoyle/netsquares/util/Maps.kt
package gargoyle.netsquares.util

import kotlin.reflect.KClass

object Maps {
    fun <K : Any, V : Any> keyForValue(map: Map<K, V>, value: V): K? =
        map.filter { (_, value1) -> value1 == value }.map { (key, _) -> key }.firstOrNull()

    fun <K : Any, V : Any> keyForValueMultiMap(map: Map<K, List<V>>, value: V): K? =
        map.filter { (_, value1) -> value1.contains(value) }.map { (key, _) -> key }.firstOrNull()

    fun <K : Any, V : Any> keyForValueMultiMapset(map: Map<K, MutableSet<V>>, value: V): K? =
        map.filter { (_, value1) -> value1.contains(value) }.map { (key, _) -> key }.firstOrNull()

    fun <K : Any?, V : Any> keyForValueMultiMapsetAny(map: Map<K, MutableSet<V>>, value: V): K? =
        map.filter { (_, value1) -> value1.contains(value) }.map { (key, _) -> key }.firstOrNull()

    fun <K : Any, V : Any> keysForValue(map: Map<K, V>, value: V): List<K> =
        map.filter { (_, value1) -> value1 == value }.map { (key, _) -> key }

    fun <K : Any, V : Any> keysForValueMultiMap(map: Map<K, List<V>>, value: V): List<K> =
        map.filter { (_, value1) -> value1.contains(value) }.map { (key, _) -> key }

    fun <K : Any, V : Any> keysForValueMultiMapset(map: Map<K, Set<V>>, value: V): List<K> =
        map.filter { (_, value1) -> value1.contains(value) }.map { (key, _) -> key }

    fun <K : Any, V : Any> map(vararg data: Any): Map<K, V> = fill(mutableMapOf(), *data)

    fun <K : Any, V : Any> mapFrom(source: () -> MutableMap<K, V>, vararg data: Any): Map<K, V> = fill(source(), *data)

    @Suppress("UNCHECKED_CAST")
    private fun <K : Any, V : Any> fill(map: MutableMap<K, V>, vararg data: Any): Map<K, V> {
        var i = 0
        while (i < data.size) {
            map[data[i] as K] = data[i + 1] as V
            i += 2
        }
        return map
    }

    fun <K : Any, V : Any> multiMap(keyType: KClass<K>, valueType: KClass<V>, vararg data: Any): Map<K, MutableList<V>> =
        multiMapFrom(keyType, valueType, { mutableMapOf() }, { mutableListOf() }, *data)

    @Suppress("UNCHECKED_CAST")
    private fun <K : Any, V : Any> multiMapFrom(
        keyType: KClass<K>,
        valueType: KClass<V>,
        mapSource: () -> MutableMap<K, MutableList<V>>,
        listSource: () -> MutableList<V>,
        vararg data: Any
    ): Map<K, MutableList<V>> {
        val map = mapSource()
        lateinit var key: K
        for (el in data) {
            if (keyType.isInstance(el)) {
                key = el as K
            } else if (valueType.isInstance(el)) {
                vs(map, key, listSource).add(el as V)
            } else {
                throw IllegalArgumentException()
            }
        }
        return map
    }

    private fun <K : Any, V : Any> vs(
        map: MutableMap<K, MutableList<V>>, key: K, listSource: () -> MutableList<V>
    ): MutableList<V> {
        return if (map[key] == null) {
            val newValue = listSource()
            map[key] = newValue
            newValue
        } else map[key]!!
    }

    private fun <K : Any, V : Any> vs(
        map: MutableMap<K, MutableSet<V>>, key: K, listSource: () -> MutableSet<V>
    ): MutableSet<V> {
        return if (map[key] == null) {
            val newValue = listSource()
            map[key] = newValue
            newValue
        } else map[key]!!
    }

    fun <K : Any, V : Any> multiMapAdd(map: MutableMap<K, MutableList<V>>, key: K, value: V): Map<K, MutableList<V>> =
        multiMapAdd({ mutableListOf() }, map, key, value)

    private fun <K : Any, V : Any> multiMapAdd(
        listSource: () -> MutableList<V>, map: MutableMap<K, MutableList<V>>, key: K, value: V
    ): Map<K, MutableList<V>> {
        vs(map, key) { listSource() }.add(value)
        return map
    }

    fun <K : Any, V : Any> multiMapAddAll(
        map: MutableMap<K, MutableList<V>>, key: K, vararg values: V
    ): Map<K, MutableList<V>> = multiMapAddAll({ mutableListOf() }, map, key, *values)

    private fun <K : Any, V : Any> multiMapAddAll(
        listSource: () -> MutableList<V>, map: MutableMap<K, MutableList<V>>, key: K, vararg values: V
    ): Map<K, MutableList<V>> {
        vs(map, key) { listSource() }.addAll(listOf(*values))
        return map
    }

    fun <K : Any, V : Any> multiMapset(
        keyType: KClass<K>, valueType: KClass<V>, vararg data: Any
    ): Map<K, MutableSet<V>> = multiMapsetFrom(keyType, valueType, { mutableMapOf() }, { mutableSetOf() }, *data)

    @Suppress("UNCHECKED_CAST")
    private fun <K : Any, V : Any> multiMapsetFrom(
        keyType: KClass<K>,
        valueType: KClass<V>,
        mapSource: () -> MutableMap<K, MutableSet<V>>,
        listSource: () -> MutableSet<V>,
        vararg data: Any
    ): Map<K, MutableSet<V>> {
        val map = mapSource()
        lateinit var key: K
        for (el in data) {
            if (keyType.isInstance(el)) {
                key = el as K
            } else if (valueType.isInstance(el)) {
                vs(map, key) { listSource() }.add(el as V)
            } else {
                throw IllegalArgumentException()
            }
        }
        return map
    }

    fun <K : Any, V : Any> multiMapsetAdd(map: MutableMap<K, MutableSet<V>>, key: K, value: V): Map<K, MutableSet<V>> =
        multiMapsetAdd({ mutableSetOf() }, map, key, value)

    private fun <K : Any, V : Any> multiMapsetAdd(
        listSource: () -> MutableSet<V>, map: MutableMap<K, MutableSet<V>>, key: K, value: V
    ): Map<K, MutableSet<V>> {
        vs(map, key) { listSource() }.add(value)
        return map
    }

    fun <K : Any, V : Any> multiMapsetAddAll(
        map: MutableMap<K, MutableSet<V>>, key: K, vararg values: V
    ): Map<K, MutableSet<V>> = multiMapsetAddAll({ mutableSetOf() }, map, key, *values)

    private fun <K : Any, V : Any> multiMapsetAddAll(
        listSource: () -> MutableSet<V>, map: MutableMap<K, MutableSet<V>>, key: K, vararg values: V
    ): Map<K, MutableSet<V>> {
        vs(map, key) { listSource() }.addAll(listOf(*values))
        return map
    }

    fun <S, T> transform(source: List<S>, mapper: (S) -> T): List<T> = source.map(mapper)
}
